website/Contribution/CONTRIBUTING.rst

```diff
@@ -1066,6 +1066,181 @@ test cases.
+      partner = self.env["res.partner"].create({"name": "Foo"})
+      self.assertEqual(partner.create_date, datetime(2024, 1, 1, 10, 10, 10))
+
+Avoid contacting external services
+----------------------------------
+
+If your module connects with some kind of external service, you should mock its
+tests to avoid false negatives when the service is down or slow. You can help
+yourself with the `unittest.mock
+<https://docs.python.org/3/library/unittest.mock.html>`_ library. Since Odoo
+16.0, it is used behind the scenes when you call ``cls.classPatch()``,
+``cls.startClassPatcher()``, ``self.patch()`` (also available in older
+versions) or ``self.startPatcher()`` in a test case:
+
+.. code-block:: python
+
+  from odoo.tests import TransactionCase
+
+  class TestSomething(TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+      super().setUpClass()
+      cls.classPatch(
+        self.env["res.partner"], "patched_method", lambda self: "Some response"
+      )
+
+    def test_something(self):
+      ...  # Your test here
+
+Example with ``unittest.mock`` directly:
+
+.. code-block:: python
+
+  from unittest.mock import patch
+  from odoo.tests import TransactionCase
+
+  class TestSomething(TransactionCase):
+    @patch("my_module.external_service")
+    def test_something(self, mock_external_service):
+      mock_external_service.return_value = "Some response"
+      # Your test here
+
+The `vcrpy <https://pypi.org/project/vcrpy-unittest/>`_ library is also
+helpful for recording and replaying HTTP interactions.
+
+Remember that what you want to test is your code, not the external service.
+There are monitoring tools for that. Also remember that, if there is a bug in
+the external service, it is not your responsibility to fix it. And if you want
+to contemplate that possibility, you just have to test the buggy behavior in a
+deterministic manner, also using mocks.
+
+Still, sometimes you may still want to test the real service, just to make sure
+they didn't change their API. In this case, `do just like upstream Odoo does
+<https://github.com/odoo/odoo/blob/226b5645c763936986a67872c7d624681cc9b9b8/addons/account/tests/test_download_xsds.py#L4>`__ and mark your test with these standard test tags:
+
+* ``-standard`` to skip these tests by default.
+* ``external`` to mark them as depending on connection with the outside world,
+  whose state is not under your control.
+* ``external_l10n``, just like ``external`` but for localization tests.
+
+You will be able then to isolate those test runs separately::
+
+  odoo -i my_module -d my_database -u my_module --test-tags=external,external_l10n
+
+Avoid relying on demo data
+--------------------------
+
+Demo data can be altered manually or by other third party module (through its
+fully XML-ID identifier). Uncontrolled inputs can produce false positives or
+negatives in your tests.
+
+Instead, create the test data within your test suite. You will have more
+consistent and resilient results.
```

Comment:
```suggestion
If you create the test data within your test suite, you will have more consistent and resilient results.
```

reformulation so that it's advice, not an obligation.